src/internet/spotifywebapi/spotifywebapiservice.cpp

```diff
@@ -0,0 +1,189 @@
+/* This file is part of Clementine.
+   Copyright 2021, Kenman Tsang <kentsangkm@pm.me>
+
+   Clementine is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   Clementine is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with Clementine.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "spotifywebapiservice.h"
+
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QtDebug>
+#include <utility>
+
+#include "3rdparty/qtiocompressor/qtiocompressor.h"
+#include "core/application.h"
+#include "core/network.h"
+#include "globalsearch/globalsearch.h"
+#include "globalsearch/spotifywebapisearchprovider.h"
+#include "ui/iconloader.h"
+
+static constexpr auto kServiceName = "SpotifyWebApi";
+
+namespace {
+template <typename T>
```

Comment:
I sort of get what these are doing but it's not the most obvious solution. Consider something more straightforward or at least documenting what these are doing.

Reply:
I am new to the Qt stuff, can you give me some hints on how it can be easier? thank you

Reply:
An iterative solution without templates would be a lot more readable, for instance.